### PR TITLE
terminate after last ipc connection closes

### PR DESCRIPTION
**Pull request by neonphog**

### CODE REVIEW

File: lib/n3h-ipc/n3hMode.js [hunk 1]
```diff
@@ -164,8 +165,22 @@ class N3hMode extends AsyncClass {
         }
         break
       case 'close':
+        if (typeof this._trackConsForTerminate !== 'number') {
+          throw new Error('hey? close without connect?')
+        } else {
+          --this._trackConsForTerminate
+        }
+        if (this._trackConsForTerminate === 0) {
```
Comment (ddd-mtl) [inline lib/n3h-ipc/n3hMode.js]: `this._trackConsForTerminate <== 0`
Seems a bit safer :)

**Comment by ddd-mtl** [comment]
@lucksus Yep. Ideally since Conductor creates the subprocess, it should control its destruction. A workaround with this would be to have conductor create a dummy IPC connection that it controls so n3h stays alive.

**Comment by neonphog** [comment]
Thanks for checking that @lucksus ! Yeah, it's a tricky situation, n3h currently only has knowledge of core's continued existence through the ipc connections. As @ddd-mtl mentions, if conductor were to open a "keep-alive" connection that sent ping requests say every 10 seconds or something, it wouldn't be that much overhead, and should be fairly straight forward to implement given the existing websocket code in the net crate.

Let us know what you think, I'm happy to write the keep-alive module.

**Comment by zippy** [comment]
https://github.com/holochain/holochain-rust/pull/1582  now has the keep-alive.  Does it need to actually send the ping requests?

**Comment by neonphog** [comment]
> holochain/holochain-rust#1582 now has the keep-alive. Does it need to actually send the ping requests?

n3h is sending ping requests to open connections, since 1582 was implemented using p2pnetwork, it should auto-respond to those ping requests, so I think we're good :+1:

**Pull request merged**